src/librbd/cache/rwl/Types.cc

```diff
@@ -79,16 +79,17 @@ ExtentsSummary<ExtentsType>::ExtentsSummary(const ExtentsType &extents)
    * and last_image_byte, inclusive, but we don't guarantee here
    * that they address all of those bytes. There may be gaps. */
   first_image_byte = extents.front().first;
-  last_image_byte = first_image_byte + extents.front().second;
+  last_image_byte = first_image_byte + extents.front().second - 1;
   for (auto &extent : extents) {
```

Comment:
nit, could use structured binding here to extract the first element and the second element out of the `std::pair<>`. as probably, they deserve better names. something like:
```c++
if (extents.empty()) return;

first_image_byte = numeric_limits<uint64_t>::max();
for (auto [offset, len] : extents) {
  first_image_byte = std::min(first_image_byte, offset);
  last_image_byte = std::max(last_image_byte, offset + len - 1);
  total_bytes += len;
}
```

Reply:
after reading through the related code, probably we should change the semantics of `BlockExtent::block_end` so a `BlockExtent` define the range of `[block_start, block_end)` instead of `[block_start, block_end]`, to avoid the off-by-one bugs in future?

Reply:
@tchaikov I'll check it this point.
By the way. The class ExtentsSummary class is used to covert some Extents into BlockExtent. Do you think we could change the class name to be ExtentsToBlockExtent?

Reply:
personally, i feel that `BlockExtent` is good enough. but Jason has the call, i believe.

Reply:
> after reading through the related code, probably we should change the semantics of `BlockExtent::block_end` so a `BlockExtent` define the range of `[block_start, block_end)` instead of `[block_start, block_end]`, to avoid the off-by-one bugs in future?

Yes, that was the original intent behind `BlockGuard` since it was added for the object map and all its operations are `[object_start, object_end)`

Reply:
Right. And also you can check the logic of BlockGuard from the test case: https://github.com/ceph/ceph/blob/master/src/test/librbd/test_BlockGuard.cc